packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/layout/layoutHandles.ts

```diff
@@ -0,0 +1,66 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Obeo.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+import { Diagram } from '../DiagramRenderer.types';
+import { getEdgeParameters } from '../edge/EdgeLayout';
+import { ConnectionHandle } from '../handles/ConnectionHandles.types';
+
+export const layoutHandles = (diagram: Diagram) => {
+  diagram.edges.forEach((edge) => {
+    const { sourceNode, targetNode, sourceHandle, targetHandle } = edge;
+    if (sourceNode && targetNode && sourceHandle && targetHandle) {
+      const { sourcePosition, targetPosition } = getEdgeParameters(sourceNode, targetNode, diagram.nodes);
+
+      const nodeSourceConnectionHandle: ConnectionHandle | undefined = sourceNode.data.connectionHandles.find(
+        (connectionHandle: ConnectionHandle) => connectionHandle.id === sourceHandle
+      );
+      const nodeTargetConnectionHandle: ConnectionHandle | undefined = targetNode.data.connectionHandles.find(
+        (connectionHandle: ConnectionHandle) => connectionHandle.id === targetHandle
+      );
+
+      if (
+        nodeSourceConnectionHandle?.position !== sourcePosition &&
+        nodeTargetConnectionHandle?.position !== targetPosition
+      ) {
+        const nodeSourceConnectionHandles: ConnectionHandle[] = sourceNode.data.connectionHandles.map(
+          (nodeConnectionHandle: ConnectionHandle) => {
+            if (nodeConnectionHandle.id === sourceHandle && nodeConnectionHandle.type === 'source') {
+              nodeConnectionHandle.position = sourcePosition;
+            }
+            return nodeConnectionHandle;
+          }
+        );
+
+        const nodeTargetConnectionHandles: ConnectionHandle[] = targetNode.data.connectionHandles.map(
+          (nodeConnectionHandle: ConnectionHandle) => {
+            if (nodeConnectionHandle.id === targetHandle && nodeConnectionHandle.type === 'target') {
+              nodeConnectionHandle.position = targetPosition;
+            }
+            return nodeConnectionHandle;
+          }
+        );
+
+        diagram.nodes = diagram.nodes.map((node) => {
+          if (edge.sourceNode && edge.targetNode) {
+            if (edge.sourceNode.id === node.id) {
+              node.data = { ...node.data, connectionHandles: nodeSourceConnectionHandles };
+            }
+            if (edge.targetNode.id === node.id) {
+              node.data = { ...node.data, connectionHandles: nodeTargetConnectionHandles };
+            }
+          }
```

Comment:
Yes, this could be improved but we only call this is the handles position changed.
`      if (
        nodeSourceConnectionHandle?.position !== sourcePosition &&
        nodeTargetConnectionHandle?.position !== targetPosition
      ) {
`